### DISCUSSION TITLE
How to detect if USB is connected ?

### DISCUSSION THREAD

**Discussion by zcattacz** [micropython/micropython#11596] [ESP32]
Hi, Is there any way to know if USB is connected or maybe a REPL session is started (whatever fits here ..., you got the idea)?  
The use case: for example, in `boot.py`, on start, if USB is connected, the default task will be skipped. If not, default task will run immediately. or user project can skip debug output based on USB connection status etc.

**Comment by jimmo** [comment]
On stm32 you can use `pyb.USB_VCP().isconnected()` to tell if something is connected. But in this case, "connected" means something has the serial port open. So I'm not quite sure that's what you mean.

But you're asking about ESP32, which doesn't have USB, rather the boards typically provide a USB to UART adaptor. I guess you might be able to sense the VBUS line? But that won't tell you whether there's a PC on the other end, or just e.g. a USB power supply.

On ESP32-S2/S3 there is USB support, but we don't currently expose any APIs for it.

**Reply by zcattacz** [reply]
Hi, thank you for the tips. I have both ESP32 and ESP32 S2,
On ESP32 S2  Is the REPL connection from USB a `serial port`, like vcom on PC ?
On ESP32, is possible to determine if the serial port is open ? I guess trying to open an opened port wouldn't raise error like on PC?
Please forgive my ignorance, not really sure what I am talking about, I know they are not linux boxes :-D

**Reply by karfas** [reply]
> Is there any way to know if USB is connected or maybe a REPL session is started

At least for the non-USB ESP32 variants, I assume a REPL session gets started always, regardless of a connected PC.
It might be possible to check the serial handshake lines (CTS, DTR, DCD?) of the UART, but this depends upon the connection between the USB/UART converter to the ESP32 **on your specific board**.

**Comment by zcattacz** [comment]
Thank you. I settled with a jumper cap.
In my case, connecting a usb cables to ESP32-S2 make the chip nonoperational, which is a deal breaker. ESP32 doesn't have this issue.
